File: TemaProiect/motoProjectC#/motoProjectCSharp/motoProjectCSharp/repos/TeamRepo.cs
namespace motoProjectCSharp.repos;

using System.Collections.Generic;
using motoProjectCSharp.domain;
using Microsoft.Data.SqlClient;
using System.Linq;
using log4net;

public class TeamRepo : ITeamRepo
    {
        private readonly string _url;
        private static readonly ILog Logger = LogManager.GetLogger(typeof(TeamRepo));

        public TeamRepo(string url)
        {
            _url = url;
        }

        public void Save(Team team)
        {
            string query = "INSERT INTO teams (name) VALUES (@name)";
            try
            {
                using (SqlConnection conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(query, conn))
                    {
                        Logger.Info("Saving team: {team.Name}");
                        cmd.Parameters.AddWithValue("@name", team.Name);
                        cmd.ExecuteNonQuery();
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error("Error connecting: {e.Message}", e);
            }
        }

        public Team? FindById(long id)
        {
            string query1 = "SELECT * FROM teams WHERE id = @id";
            string query2 = "SELECT * FROM teams_participants WHERE team_id = @team_id";
            string query3 = "SELECT * FROM participants WHERE id IN ({0})";

            long? teamId = null;
            string? teamName = null;
            try
            {
                using (SqlConnection conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(query1, conn))
                    {
                        Logger.Info("Finding team by id: {id}");
                        cmd.Parameters.AddWithValue("@id", id);
                        using (SqlDataReader reader = cmd.ExecuteReader())
                        {
                            if (reader.Read())
                            {
                                teamId = reader.GetInt64(reader.GetOrdinal("id"));
                                teamName = reader.GetString(reader.GetOrdinal("name"));
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error("Error connecting: {e.Message}", e);
            }

            if (!teamId.HasValue)
            {
                Logger.Info("Team not found: {id}");
                return null;
            }

            List<long> memberIds = new List<long>();
            try
            {
                using (SqlConnection conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(query2, conn))
                    {
                        Logger.Info("Finding members by team id: {id}");
                        cmd.Parameters.AddWithValue("@team_id", id);
                        using (SqlDataReader reader = cmd.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                long memberId = reader.GetInt64(reader.GetOrdinal("participant_id"));
                                memberIds.Add(memberId);
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error("Error connecting: {e.Message}", e);
            }

            List<Participant> members = new List<Participant>();
            string placeholders = string.Join(",", memberIds.Select((memberId, index) => "@id" + index));
            string query3Formatted = string.Format(query3, placeholders);

            try
            {
                using (SqlConnection conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(query3Formatted, conn))
                    {
                        Logger.Info("Finding members in list");
                        for (int i = 0; i < memberIds.Count; i++)
                        {
                            cmd.Parameters.AddWithValue("@id" + i, memberIds[i]);
                        }
                        using (SqlDataReader reader = cmd.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                long memberId = reader.GetInt64(reader.GetOrdinal("id"));
                                string memberName = reader.GetString(reader.GetOrdinal("name"));
                                string memberIdNumber = reader.GetString(reader.GetOrdinal("id_number"));
                                long memberEngineSize = reader.GetInt64(reader.GetOrdinal("engine_size"));
                                members.Add(new Participant(memberId, memberName, memberIdNumber, memberEngineSize));
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error("Error connecting: {e.Message}", e);
            }

            Team team = new Team(teamId.Value, teamName ?? "default", members);
            return team;
        }

        public List<Team> FindAll()
        {
            string query = "SELECT * FROM teams";
            List<Team> teams = new List<Team>();

            try
            {
                using (SqlConnection conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(query, conn))
                    {
                        Logger.Info("Finding all teams");
                        using (SqlDataReader reader = cmd.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                long teamId = reader.GetInt64(reader.GetOrdinal("id"));
                                Team? team = FindById(teamId);
                                if (team != null)
                                {
                                    teams.Add(team);
                                }
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error("Error connecting: {e.Message}", e);
            }

            return teams;
        }
    }

File: TemaProiect/motoProject/src/main/java/org/example/motoproject/repos/TeamRepo.java
package org.example.motoproject.repos;

import org.example.motoproject.domain.Participant;
import org.example.motoproject.domain.Team;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class TeamRepo implements TeamRepoInterface {
    private final String url;
    private static final Logger logger = LogManager.getLogger(TeamRepo.class);

    public TeamRepo(String url) {
        this.url = url;
    }

    @Override
    public void save(Team team) {
        String query = "INSERT INTO teams (name) VALUES (?)";
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query)) {
            logger.info("Saving team: {}", team);
            stmt.setString(1, team.getName());
            stmt.executeUpdate();
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
    }

    @Override
    public Optional<Team> findById(Long id) {
        String query1 = "SELECT * FROM teams WHERE id = ?";
        String query2 = "SELECT * form teams_participants WHERE team_id = ?";

        Long teamId = null;
        String teamName = null;
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query1)) {
            logger.info("Finding team by id: {}", id);
            stmt.setLong(1, id);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                teamId = rs.getLong("id");
                teamName = rs.getString("name");
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }

        if (teamId == null) {
            logger.info("Team not found: {}", id);
            return Optional.empty();
        }

        List<Long> memberIds = new ArrayList<>();
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query2)) {
            logger.info("Finding members by id: {}", id);
            stmt.setLong(1, id);
            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                Long memberId = rs.getLong("participant_id");
                memberIds.add(memberId);
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }

        List<Participant> members = new ArrayList<>();
        String placeholders = String.join(",", Collections.nCopies(memberIds.size(), "?"));
        String query3 = "SELECT * FROM participants WHERE id IN (" + placeholders + ")";
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query3)) {
            logger.info("Finding members in list");
            for (int i = 0; i < memberIds.size(); i++) {
                stmt.setLong(i + 1, memberIds.get(i));
            }
            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                Long memberId = rs.getLong("id");
                String memberName = rs.getString("name");
                String memberIdNumber = rs.getString("id_number");
                Long memberEngineSize = rs.getLong("engine_size");
                members.add(new Participant(memberId, memberName, memberIdNumber, memberEngineSize));
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
        Team team = new Team(teamId, teamName, members);
        return Optional.of(team);
    }

    @Override
    public List<Team> findAll() {
        String query = "SELECT * FROM teams";
        List<Team> teams = new ArrayList<>();
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query)) {
            logger.info("Finding all teams");
            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                Long teamId = rs.getLong("id");
                findById(teamId).ifPresent(teams::add);
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
        return teams;
    }
}
